bokehjs/src/lib/document/document.ts

```diff
@@ -703,8 +703,20 @@ export class Document {
     for (const event_json of events_json) {
       switch (event_json.kind) {
         case 'MessageSent': {
-          const msg_data = Document._resolve_refs(event_json.msg_data, old_references, new_references)
-          this._trigger_on_message(event_json.msg_type, msg_data)
+          const {msg_type, msg_data} = event_json
+          let data: unknown
+          if (msg_data === undefined) {
+            if (buffers.length == 1) {
+              const [[, buffer]] = buffers
+              data = buffer
+            } else {
+              throw new Error("expected exactly one buffer")
+            }
+          } else {
+            data = Document._resolve_refs(msg_data, old_references, new_references)
+          }
```

Comment:
This (switching on implicitly defined flags to find buffers that are sometimes there, sometimes not) seems like the sort of thing you usually disfavor, is there not a cleaner way you'd rather do things here?

Reply:
This is not a flag. You have either `msg_data` and no buffer or no `msg_data` and a buffer, resulting in `data` being an object (typically JSON string for case specific deserialization) or `ArrayBuffer` (for case specific decoding). There is a better approach, but that will be a part of bokeh 3.0.